ydb/mvp/oidc_proxy/oidc_protected_page_nebius.cpp

```diff
@@ -105,6 +123,43 @@ void THandlerSessionServiceCheckNebius::ExchangeSessionToken(const TString sessi
     Become(&THandlerSessionServiceCheckNebius::StateExchange);
 }
 
+void THandlerSessionServiceCheckNebius::ExchangeImpersonatedToken(const TString sessionToken, const TString impersonatedToken, const NActors::TActorContext& ctx) {
+    LOG_DEBUG_S(ctx, EService::MVP, "Exchange impersonated token");
+    tokenExchangeType = ETokenExchangeType::ImpersonatedToken;
+    NHttp::THttpOutgoingRequestPtr httpRequest = NHttp::THttpOutgoingRequest::CreateRequestPost(Settings.GetExchangeEndpointURL());
+    httpRequest->Set<&NHttp::THttpRequest::ContentType>("application/x-www-form-urlencoded");
+
+    TMvpTokenator* tokenator = MVPAppData()->Tokenator;
+    TString token = "";
+    if (tokenator) {
+        token = tokenator->GetToken(Settings.SessionServiceTokenName);
+    }
+    httpRequest->Set("Authorization", token); // Bearer included
+    TStringBuilder body;
+    body << "grant_type=urn:ietf:params:oauth:grant-type:token-exchange"
+            << "&requested_token_type=urn:ietf:params:oauth:token-type:access_token"
+            << "&subject_token_type=urn:ietf:params:oauth:token-type:jwt"
+            << "&subject_token=" << impersonatedToken
+            << "&actor_token=" << sessionToken
+            << "&actor_token_type=urn:ietf:params:oauth:token-type:session_token";
+    httpRequest->Set<&NHttp::THttpRequest::Body>(body);
+
+    ctx.Send(HttpProxyId, new NHttp::TEvHttpProxy::TEvHttpOutgoingRequest(httpRequest));
+
+    Become(&THandlerSessionServiceCheckNebius::StateExchange);
+}
+
+void THandlerSessionServiceCheckNebius::ClearImpersonatedCookie(const NActors::TActorContext& ctx) {
+    TString impersonatedCookieName = CreateNameImpersonatedCookie(Settings.ClientId);
+    LOG_DEBUG_S(ctx, EService::MVP, "Clear impersonated cookie (" << impersonatedCookieName << ") and retry");
+    NHttp::THeadersBuilder responseHeaders;
+    SetCORS(Request, &responseHeaders);
+    responseHeaders.Set("Set-Cookie", ClearSecureCookie(impersonatedCookieName));
+    responseHeaders.Set("Location", Request->URL);
+    ctx.Send(Sender, new NHttp::TEvHttpProxy::TEvHttpOutgoingResponse(Request->CreateResponse("307", "Temporary Redirect", responseHeaders)));
+    Die(ctx);
+}
```

Comment:
There is similar code in `ydb/mvp/oidc_proxy/oidc_impersonate_stop_page_nebius.cpp`

Reply:
I would prefer to leave it as is. The difference in logging and headers means that refactoring would only save about 3-4 lines, which seems negligible.